requirements.txt

```diff
@@ -1,2 +1,3 @@
 pip<24.0
 immutabledict<2.2.5
+typing_extensions>=4.6,<5
```

Comment:
what do you need this for? If it's for Self then I'd recommend using `from __future__ import annotations` instead, to avoid the extra dependency

Reply:
It is used in a ```@classmethod``` with return type ```Self```. The function will likely be overwritten by inherited class and the return type should be that inherited class instead of the base class.

Moreover, it will be easier for migration if we are going to drop old python supports as they reach end of life.

Reply:
[Self in classmethod](https://peps.python.org/pep-0673/#use-in-classmethod-signatures)

We can use a "workaround" like ```Self = TypeVar("Self", bound="Shape")``` but it will lead to a namespace pollution.
For example [this](https://github.com/xdslproject/xdsl/blob/bc047cc69cdae6030e931b0bc48c57dee374760d/xdsl/ir/core.py#L355) is indeed a workaround but it is no longer used in current version of xdsl (as far as I have checked). It is a left behind type variable which can be traced back to xdsl v0.4.0 (no longer on github but I did use it in CT course 2 years ago.) It is common for people to use ```from some_lib import *``` and they will find it confusing about where the ```A``` comes from.